crates/matrix-sdk/src/room/mod.rs

```diff
@@ -403,6 +403,20 @@ impl Room {
         IdentityStatusChanges::create_stream(self.clone()).await
     }
 
+    /// Mark this user as "pinned" meaning that we have seen their cryptographic
+    /// identity before.
+    #[cfg(feature = "e2e-encryption")]
+    pub async fn pin_user_identity(&self, user_id: &UserId) -> Result<()> {
```

Comment:
We're not saying *which* identity to pin, only which user's identity should be pinned, so this feels a bit smelly. I'm not sure how we will be using this API, but is there potential for a switcheroo/race here?

We observe one identity for some user U, which we want to pin. But then it switches under us, which we don't notice because the API is simply "pin *some* identity for user U, whichever one you find".

Reply:
Even if we are sure that a switcheroo can't happen in a dangerous manner at this point in time, and in the way EX wants to use it right now, things might change in the future and we have other users of the bindings. We can't guarantee that you can't shoot yourself in the foot this way.

There is a reason why this isn't exposed on the room in the main crate and the API for this should be kept 1:1.

Reply:
Sorry, unresolving, since I didn't see poljar's comment before hitting resolve (the UI didn't update).

Reply:
Hmm, good point. I've re-structured the interface and force-pushed. Now we can get the identity first, and pin it later.

We don't currently have a good way of actually identifying which identity we have, but I think that is outside the scope of this change. (I suspect we should expose the master key?)

Reply:
Sry @poljar I am a bit lost. Why would keeping the API 1:1 will help for the "switcheroo" thing?
If I understand well, what would fix it is that the `pin` method takes a "master_key" as parameter and that the room identity state changes stream gives you the `master_key` to confirm that you can then later pass back as a parameter.

But even that doesn't really help the race scenario in case of 2 consecutive changes? Not sure the current pinning mecanism works like that.

Reply:
> If I understand well, what would fix it is that the pin method takes a "master_key" as parameter 

The `pin()` method takes the master key as an argument, it's just that the master key is inside of `self.inner.master_key`.

So yes, you are correct, the master key needs to be part of the function call.

> But even that doesn't really help the race scenario in case of 2 consecutive changes? Not sure the current pinning mecanism works like that.

The changes can happen at any point in time, but you have a copy of the identity which will not change under your feet while you decide if you want to pin it or not.

Reply:
@poljar does the new version match what you expected?